src/cmap/connection.ts

```diff
@@ -246,10 +247,22 @@ export class Connection extends TypedEventEmitter<ConnectionEvents> {
     this[kIsMaster] = response;
   }
 
+  get serviceId(): ObjectId | undefined {
+    return this.ismaster?.serviceId;
+  }
+
+  get loadBalanced(): boolean | undefined {
+    return this.description.loadBalanced;
+  }
+
   get generation(): number {
     return this[kGeneration] || 0;
   }
 
+  set generation(generation: number) {
```

Comment:
That's a tricky one. In order to maintain that, we need to know the generation when the connection is instantiated. In order to know that, we need to know the `serviceId` - which requires us to perform the initial handshake.

Reply:
I see, I think at very least we should make this a very explicitly named method then rather than a mutator (maybe `updateGenerationForServiceId`) to make the intent very clear. 

I'm still pondering whether we can do this while making `generation` immutable.. A `Connection` is created in the `connect` method, and the `generation` passed here is provided by the `ConnectionPool` when calling `connect`. Since your service generation authority is a map, perhaps you can pass a reference to that into the `connect` options, and resolve the correct generation there?
```
performHandshake(..., (err, hello) => {
  if (hello.serviceId && options.serviceGenerations.has(hello.serviceId)) {
     // modify the generation of the options passed into the `Connection` constructor
  }
});
```

It seems like a big contortion, but I think it's important for avoiding subtle bugs that we maintain immutability where it has been introduced (to solve bugs in the past!)

Reply:
Even passing in the generation we still execute the handshake after the `Connection` is instantiated, so we still don't have the `serviceId`. I'm happy to explicitly name the method to set it but I don't see any way around this unless we change the driver to execute a handshake before instantiating a `Connection`, which seems really strange to me.